js/l1_player.js

```diff
@@ -0,0 +1,258 @@
+    },
+    addTracks(list) {
+      chrome.runtime.getBackgroundPage((w) => {
+        w.player.appendAudioList(list);
+      });
+    },
+    clearPlaylist() {
+      chrome.runtime.getBackgroundPage((w) => {
+        w.player.clearPlaylist();
+      });
+    },
+    setNewPlaylist(list) {
+      chrome.runtime.getBackgroundPage((w) => {
+        w.player.setNewPlaylist(list);
+      });
+    },
+    getTrackById(id) {
+      if (!l1Player.status.playlist) return null;
+      return l1Player.status.playlist.find(track => track.id === id);
+    },
+    setBootstrapTrack(fn) {
+      l1Player.bootstrapTrack = fn;
+    },
+    connectPlayer() {
+      const player = this;
+      chrome.runtime.getBackgroundPage((w) => {
+        w.player.sendFullUpdate();
+        w.player.sendPlaylistEvent();
+        w.player.sendPlayingEvent();
+        w.player.sendLoadEvent();
+        if (!w.player.playing) {
+          initPlayer(player);
+        }
+      });
+    },
+  };
+
+  l1Player.injectDirectives = (ngApp) => {
+    ngApp.directive('playFromPlaylist', () => ({
+      restrict: 'EA',
+      scope: {
+        song: '=playFromPlaylist',
+      },
+      link(scope, element) {
+        element.bind('click', () => {
+          l1Player.playById(scope.song.id);
+        });
+      },
+    }));
+
+    ngApp.directive('nextTrack', () => ({
+      restrict: 'EA',
+      link(scope, element) {
+        element.bind('click', () => {
+          l1Player.next();
+        });
+      },
+    }));
+
+    ngApp.directive('prevTrack', () => ({
+      restrict: 'EA',
+      link(scope, element) {
+        element.bind('click', () => {
+          l1Player.prev();
+        });
+      },
+    }));
+
+    ngApp.directive('clearPlaylist', () => ({
+      restrict: 'EA',
+      link(scope, element) {
+        element.bind('click', () => {
+          l1Player.clearPlaylist();
+        });
+      },
+    }));
+
+    ngApp.directive('playPauseToggle', () => ({
+      restrict: 'EA',
+      link(scope, element) {
+        element.bind('click', () => {
+          l1Player.toggleplayPause();
+        });
+      },
+    }));
+  };
+
+  chrome.runtime.onMessage.addListener((msg, sender, res) => {
+    if (msg.type === 'BG_PLAYER:FULL_UPDATE') {
+      l1Player.status = {
+        ...l1Player.status,
+        ...msg.data,
+      };
+    }
+    if (msg.type === 'BG_PLAYER:FRAME_UPDATE') {
+      l1Player.status.playing = {
+        ...l1Player.status.playing,
+        ...msg.data,
+      };
+    }
+    if (msg.type === 'BG_PLAYER:PLAYLIST') {
+      l1Player.status.playlist = msg.data || [];
+    }
+    if (msg.type === 'BG_PLAYER:RETRIEVE_URL') {
+      if (l1Player.bootstrapTrack) {
+        l1Player.bootstrapTrack({
+          /**
+           * A mock sound object to set url back in player.
+           * @param {string} val
+           */
+          set url(val) {
+            chrome.runtime.getBackgroundPage((w) => {
+              w.player.setMediaURI(val, msg.data.url);
```

Comment:
w.player.setMediaURI(val, msg.data.url);

->

w.player.setMediaURI(val, msg.data.id);

Reply:
A question here. It seems the unplayable media replacement feature is using the url property to pass the media alternative. If I'm right, in the situation, the media URI should utilize the id in the url property to be retrieved. 
#320 
loweb.js line 308
```javascript
          async.parallel(
            getUrlFnList,
            function(err) {
              if(err){
                // use error to make instant return, error contains url
                sound.url = err;
                playerSuccessCallback();
                success();
                return;
              }
              playerFailCallback();
              failure();
            });
```

Reply:
A question here. It seems the unplayable media replacement feature is using the url property to pass the media alternative.

Yes, you're right. But more precisely speaking, url field stores song downloadable url and there's nothing more. 

If I'm right, in the situation, the media URI should utilize the id in the url property to be retrieved.
#320

indeed, when fail happened, url field is empty string(as I replied in last comment). there's no any info from this field, we use track object to get title and artiest field and search other platform, so you can find code like this:

       //TODO: better query method
        const keyword = track.title + ' ' + track.artist;